File: src/main/java/de/spries/fleetcommander/model/core/Player.java
package de.spries.fleetcommander.model.core;

import java.util.List;
import java.util.stream.Collectors;

import org.apache.commons.lang3.StringUtils;

import de.spries.fleetcommander.model.core.common.IllegalActionException;
import de.spries.fleetcommander.model.core.universe.FactorySite;

public class Player {

	public static class InsufficientCreditsException extends IllegalActionException {

		public InsufficientCreditsException(String msg) {
			super(msg);
		}
	}

	protected static final int STARTING_CREDITS = 500;
	protected static final int MAX_CREDITS = 99_999;

	private int id;
	private String name;
	private int credits;
	private boolean active;

	public Player(String name) {
		id = -1;
		this.name = name;
		credits = STARTING_CREDITS;
		active = true;
	}

	public int getId() {
		return id;
	}

	public void setId(int id) {
		this.id = id;
	}

	public String getName() {
		return name;
	}

	public int getCredits() {
		return credits;
	}

	public boolean isActive() {
		return active;
	}

	public boolean isHumanPlayer() {
		return true;
	}

	public void setActive(boolean active) {
		this.active = active;
	}

	public boolean canAffordFactory() {
		return credits >= FactorySite.FACTORY_COST;
	}

	public void reduceCredits(int debit) {
		if (debit > credits) {
			throw new InsufficientCreditsException("You don't have sufficient credits!");
		}
		credits -= debit;
	}

	public void addCredits(int creditsToAdd) {
		credits += creditsToAdd;
		if (credits > MAX_CREDITS) {
			credits = MAX_CREDITS;
		}
	}

	@SuppressWarnings("unused")
	public void notifyNewTurn(Game game) {
		//Nothing to do
	}

	@Override
	public String toString() {
		return name;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + id;
		result = prime * result + (name == null ? 0 : name.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (getClass() != obj.getClass()) {
			return false;
		}
		Player other = (Player) obj;
		if (id != other.id) {
			return false;
		}
		if (!StringUtils.equals(name, other.name)) {
			return false;
		}
		return true;
	}

	/**
	 * For use in tests only
	 */
	protected void setCredits(int credits) {
		this.credits = credits;
	}

	public static List<Player> filterAllOtherPlayers(List<Player> players, Player viewingPlayer) {
		return players.stream()
				.filter(p -> !p.equals(viewingPlayer)).collect(Collectors.toList());
	}
}

File: src/main/kotlin/de/spries/fleetcommander/model/core/Player.kt
package de.spries.fleetcommander.model.core

import de.spries.fleetcommander.model.core.common.IllegalActionException
import de.spries.fleetcommander.model.core.universe.FactorySite

open class Player(private val name: String,
                  private var id: Int = -1,
                  private var credits: Int = STARTING_CREDITS) {

    class InsufficientCreditsException(msg: String) : IllegalActionException(msg)

    private var status = Status.PLAYING

    fun getName() = name
    fun getId() = id
    fun getCredits() = credits
    fun getStatus() = status
    fun setCredits(credits: Int) {
        this.credits = credits
    }

    fun assignId(id: Int) {
        this.id = id
    }

    open fun isHumanPlayer(): Boolean {
        return true
    }

    fun isActive(): Boolean {
        return Status.PLAYING == status || Status.READY == status
    }

    fun isReady(): Boolean {
        return Status.READY == status
    }

    enum class Status {
        PLAYING,
        READY,
        DEFEATED,
        QUIT
    }

    fun handleDefeat() {
        status = Status.DEFEATED
    }

    fun setReady() {
        if (Status.READY == status) {
            throw IllegalActionException("You have to wait for the other players")
        }
        status = Status.READY
    }

    fun setPlaying() {
        if (Status.QUIT == status || Status.DEFEATED == status) {
            throw IllegalActionException("You're out!")
        }
        status = Status.PLAYING
    }

    fun handleQuit() {
        if (Status.QUIT == status) {
            throw IllegalActionException("You quit the game already")
        }
        status = Status.QUIT
    }

    fun canAffordFactory(): Boolean {
        return credits >= FactorySite.FACTORY_COST
    }

    fun reduceCredits(debit: Int) {
        if (debit > credits) {
            throw InsufficientCreditsException("You don't have sufficient credits!")
        }
        credits -= debit
    }

    fun addCredits(creditsToAdd: Int) {
        credits += creditsToAdd
        if (credits > MAX_CREDITS) {
            credits = MAX_CREDITS
        }
    }

    open fun notifyNewTurn(game: Game) {
        //Nothing to do
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Player

        if (name != other.name) return false

        return true
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }

    override fun toString(): String {
        return name
    }

    companion object {

        const val STARTING_CREDITS = 500
        const val MAX_CREDITS = 99999

        fun filterAllOtherPlayers(players: List<Player>, viewingPlayer: Player): List<Player> {
            return players.filter { p -> p != viewingPlayer }
        }
    }
}
